helm/openwhisk/templates/scheduler-pod.yaml

```diff
@@ -0,0 +1,273 @@
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: {{ .Release.Name }}-scheduler
+  labels:
+    name: {{ .Release.Name }}-scheduler
+{{ include "openwhisk.label_boilerplate" . | indent 4 }}
+spec:
+  serviceName: {{ .Release.Name }}-scheduler
+  podManagementPolicy: "Parallel"
+  replicas: {{ .Values.scheduler.replicaCount }}
+  selector:
+    matchLabels:
+      name: {{ .Release.Name }}-scheduler
+  template:
+    metadata:
+      labels:
+        name: {{ .Release.Name }}-scheduler
+{{ include "openwhisk.label_boilerplate" . | indent 8 }}
+
+      {{- if .Values.metrics.prometheusEnabled }}
+      annotations:
+        prometheus.io/scrape: 'true'
+        prometheus.io/port:   '{{ .Values.scheduler.port }}'
+      {{- end }}
+
+    spec:
+      serviceAccountName: {{ .Release.Name }}-core
+      restartPolicy: {{ .Values.scheduler.restartPolicy }}
+
+      {{- if .Values.affinity.enabled }}
+      affinity:
+{{ include "openwhisk.affinity.core" . | indent 8 }}
+{{ include "openwhisk.affinity.selfAntiAffinity" ( printf "%s-scheduler" .Release.Name ) | indent 8 }}
+      {{- end }}
+
+      {{- if .Values.toleration.enabled }}
+      tolerations:
+{{ include "openwhisk.toleration.core" . | indent 8 }}
+      {{- end }}
+
+      initContainers:
+      # The scheduler must wait for kafka and/or couchdb to be ready before it starts
+{{ include "openwhisk.readiness.waitForController" . | indent 6 }}
+{{ include "openwhisk.docker.imagePullSecrets" . | indent 6 }}
+
+      containers:
+      - name: scheduler
+        imagePullPolicy: {{ .Values.scheduler.imagePullPolicy | quote }}
+        image: "{{- .Values.docker.registry.name -}}{{- .Values.scheduler.imageName -}}:{{- .Values.scheduler.imageTag -}}"
+        command: ["/bin/bash", "-c", "/init.sh `hostname | awk -F '-' '{print $NF}'`"]
+        ports:
+        - name: scheduler
+          containerPort: {{ .Values.scheduler.port }}
+        - name: rpc
+          containerPort: {{ .Values.scheduler.rpcPort }}
+        - name: akka-remoting
+          containerPort: 25520
+        - name: akka-mgmt-http
+          containerPort: 19999
+        livenessProbe:
+          httpGet:
+            path: "/ping"
+            port: {{ .Values.scheduler.port }}
+            scheme: "HTTP"
+          initialDelaySeconds: {{ .Values.probes.scheduler.livenessProbe.initialDelaySeconds }}
+          periodSeconds: {{ .Values.probes.scheduler.livenessProbe.periodSeconds }}
+          timeoutSeconds: {{ .Values.probes.scheduler.livenessProbe.timeoutSeconds }}
+        readinessProbe:
+          httpGet:
+            path: "/ping"
+            port: {{ .Values.scheduler.port }}
+            scheme: "HTTP"
+          initialDelaySeconds: {{ .Values.probes.scheduler.readinessProbe.initialDelaySeconds }}
+          periodSeconds: {{ .Values.probes.scheduler.readinessProbe.periodSeconds }}
+          timeoutSeconds: {{ .Values.probes.scheduler.readinessProbe.timeoutSeconds }}
+        env:
```

Comment:
I think the `whick_cluster_name` needs to be added to differentiate different versions of clusters.
https://github.com/apache/openwhisk/blob/master/ansible/roles/schedulers/tasks/deploy.yml#L223

Reply:
Similarly, would be great to add `kafka_topic_prefix` too.
https://github.com/apache/openwhisk/blob/master/ansible/roles/schedulers/tasks/deploy.yml#L150

BTW, it seems the prefix configuration for kafka topic is not applied in this repo at all.

Reply:
The `scheduling` configurations are required.

```
      "CONFIG_whisk_scheduler_scheduling_staleThreshold": "{{ scheduler.scheduling.staleThreshold }}"
      "CONFIG_whisk_scheduler_scheduling_checkInterval": "{{ scheduler.scheduling.checkInterval }}"
      "CONFIG_whisk_scheduler_scheduling_dropInterval": "{{ scheduler.scheduling.dropInterval }}"
```
https://github.com/apache/openwhisk/blob/master/ansible/roles/schedulers/tasks/deploy.yml#L117

Reply:
`CONFIG_whisk_scheduler_queue_maxBlackboxRetentionMs` is required.
https://github.com/apache/openwhisk/blob/master/ansible/roles/schedulers/tasks/deploy.yml#L128

Reply:
Regarding ```whisk_cluster_name```: I believe you name a deployment when you create it with helm (which becomes ```Release.name``` in the helm template files), which would allow you to deploy multiple clusters without conflict (I am speaking hypothetically, as I have not tried this myself). If I am understanding correctly, perhaps ```whisk_cluster_name``` is not needed here. @dgrove-oss does this seem correct?

Reply:
The cluster name is used to differentiate multiple versions when two different releases share the same ETCD.
https://github.com/apache/openwhisk/blob/master/common/scala/src/main/scala/org/apache/openwhisk/core/etcd/EtcdUtils.scala#L72

In this change, it seems each release uses its own ETCD but when an external ETCD is shared among multiple clusters for some reason(some may want to deploy clusters in a blue-green way), the config would be required.

Reply:
I have some confusion with how ```whisk cluster name``` is implemented in the main OpenWhisk repo, and I'd appreciate your help in understanding why it is the way it is.

In the link you pasted it is ```CONFIG_whisk_cluster_name```, but in config hierarchy it is ```whisk.cluster_name```. My understanding is that it should be ```CONFIG_whisk_clusterName``` to follow convention.

To my understanding it is etcd specific, so shouldn't it be under ```etcd.clusterName``` and not in ```whisk.clusterName```? I believe elasticsearch config takes this approach here: https://github.com/apache/openwhisk/blob/a1639f0e4d7270c9a230190ac26acb61413b6bbb/ansible/roles/elasticsearch/templates/elasticsearch.yml.j2#L1

I am asking these questions because I don't want to modify the main whisk config for scheduler-specific things if I don't have to. If there is a reason for it to be the way it is currently in OpenWhisk, I can still add it to this PR. But I wanted some clarification first.

Reply:
It is not only being used by schedulers but also by controllers too.
https://github.com/apache/openwhisk/blob/1fbfd6bcb3d65845a328382641c1db7182763ced/core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/FPCPoolBalancer.scala#L72